BRANCHING.md

```diff
@@ -0,0 +1,77 @@
+- `compatible`: scheduled to be released.
+  - It contains all the changes which are literally backwards compatible with what people currently run on the mainnet. Any nodes running a version of mina based off of compatible should connect to the current mainnet.
+- `rampup`: what is deployed on the testnet
+  - Compatible with `compatible`.
+  - Never make PRs to `rampup` unless you're explicitly fixing a testnet bug.
+- `berkeley`/`izmir`: next hardfork branch / release candidate.
+  - Contains all the new features that are scheduled for the release (berkeley or izmir).
+  - On releases. Current release is "mainnet", public. Next release is "berkeley", and the one after is "izmir". The "devnet" testnet is running from `master`, sometimes `compatible`, and features only the current release (not cutting edge/berkeley).
+- `develop`: next non-harmful release (after `berkeley`).
+  - Does *not* contain cutting edge new protocol features enabled (that might go into the next release).
+  - Contains changes which break backwards compatibility, or changes that depend on past compatibility-breaking changes.  “Not backwards compatible” means that a daemon running this version of mina will not connect to mainnet.
+  - Major changes to the daemon, protocol, or crypto will sometimes cause backwards-compatibility breaking changes, and of course such changes need to be done with deliberation and are not to be taken lightly.  Changes to infrastructure, auxiliary develop scripts, tests, CI, are usually not be backwards compatibility breaking, and thereby should go into compatible (unless you are doing something very special and you know what you’re doing).
+  - The difference between `develop` and `berkeley` is that `berkeley` will be the actual hardfork release, while `develop` is subsequent softfork release candidate, softfork after `berkeley`. `develop` is just not tested as rigorously, but it's softfork compatible with `berkeley`. So if `berkeley` can be thought of as 2.0, then `develop` is 2.01.
+- Direction of merge:
+  - The back-merging direction is `master` to `compatible` to `rampup` to the next hardfork (now, `berkeley`) to `develop`.
+    - So `develop` contains all the changes from the more stable branches.
+  - Merging forward (when introducing features/fixes) is up this stack, but the place the change is introduced can be different (e.g. `compatible` for softfork features, `develop` for more experimental/next release).
+- Hard fork/Release:
+  - Whenever a hard fork happens, the code in  `develop` is released.  When this happens, the current `compatible` is entirely discarded and a new `compatible` gets created based off of `develop`
+  - So after Berkeley release: `berkeley` branch will become the new `master`. `berkeley` will be removed from `proof-systems`. `develop` will be renamed into `compatible`.
+- `o1js-main`: compatible with testnet, but has latest `proof-systems` features so that they can be used in `o1js`
```

Comment:
@mitschabaude Could you please check this? This is what I understood from Matthew, but you are saying that `o1js/main` uses `proof-systems/o1js-main` and not `proof-systems/develop`?

Reply:
yes I'm saying that but the description is still correct, at least it's the _intention_ that `o1js-main` has the latest proof-systems features so they can be used in o1js.

the only reason that this is not always true is because the crypto sometimes doesn't land the newest features in `o1js-main` 😅  and by implication, `o1js-main` then can't use `proof-systems/develop` simply because there are changes there which need adaptions in `mina`, and those adaptations haven't been landed in `o1js-main`.